indigo/indigo-extras/src/main/scala/indigoextras/ui/RadioButton.scala

```diff
@@ -0,0 +1,206 @@
+package indigoextras.ui
+
+import indigo.shared.EqualTo._
+import indigo.shared.Outcome
+import indigo.shared.datatypes.{Depth, Point, Rectangle}
+import indigo.shared.events.GlobalEvent
+import indigo.shared.input.Mouse
+import indigo.shared.scenegraph.Group
+
+/**
+  * Structure for presenting a radio button.
+  *
+  * @param buttonAssets The graphics to use identically for all of the options: up (unselected), down (selected) and over
+  * @param bounds A list of the rectangle locations of each option button by index
+  * @param depth The depth at which to present this button
+  * @param selected The index of the option currently selected, if any
+  * @param over The index of the option currently hovered over, if any
+  * @param onUnselected Events to fire when an option becomes unselected, mapping from the index of the option deselected
+  * @param onSelected Events to fire when an option becomes selected, mapping from the index of the option selected
+  * @param onHoverOver Events to fire when an option is hovered over, mapping from the index of the option hovered over
+  * @param onHoverOut Events to fire when an option is no longer hovered over, mapping from the index of the option no longer hovered over
+  */
+final case class RadioButton(
+    buttonAssets: ButtonAssets,
+    bounds: List[Rectangle],
+    depth: Depth,
+    selected: Option[Int],
+    over: Option[Int],
+    onUnselected: Int => List[GlobalEvent],
+    onSelected: Int => List[GlobalEvent],
+    onHoverOver: Int => List[GlobalEvent],
+    onHoverOut: Int => List[GlobalEvent]
+) {
+
+  /**
+    * Returns the button state of a given option: up (unselected), down (selected) or hovered over
+    *
+    * @param option The index of the option
+    * @return The state of the option's button
+    */
+  def state(option: Int): ButtonState =
+    if (selected.contains(option)) ButtonState.Down
+    else if (over.contains(option)) ButtonState.Over
+    else ButtonState.Up
+
+  /**
+    * Update all the option buttons according to the newest state of mouse input.
+    *
+    * @param mouse The current mouse state
+    * @return An Outcome with this radio button's new state
+    */
+  def update(mouse: Mouse): Outcome[RadioButton] = {
+    val nowOver: Int = bounds.indexWhere(_.isPointWithin(mouse.position))
+
+    if (nowOver >= 0) {
+      val nowSelected: Option[Int] = if (mouse.mousePressed) Some(nowOver) else selected
+      Outcome(this.copy(over = Some(nowOver), selected = nowSelected))
+        .addGlobalEvents(changeOverEvents(Some(nowOver)))
+        .addGlobalEvents(changeSelectedEvents(nowSelected))
```

Comment:
For _slightly_ more efficient construction when you're writing components or engine internals, do this instead:

```
Outcome(
  this.copy(over = Some(nowOver), selected = nowSelected),
  changeOverEvents(Some(nowOver)) ++ changeSelectedEvents(nowSelected)
)
```

Reply:
Done!